presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyAggregation.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.differentialentropy;
+
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.function.AggregationFunction;
+import com.facebook.presto.spi.function.AggregationState;
+import com.facebook.presto.spi.function.CombineFunction;
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.InputFunction;
+import com.facebook.presto.spi.function.OutputFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.type.StandardTypes;
+import io.airlift.slice.Slice;
+
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+
+@AggregationFunction("differential_entropy")
+@Description("Computes differential entropy based on random-variable samples")
+public final class DifferentialEntropyAggregation
+{
+    private DifferentialEntropyAggregation() {}
+
+    @InputFunction
+    public static void input(
+            @AggregationState State state,
+            @SqlType(StandardTypes.BIGINT) long size,
+            @SqlType(StandardTypes.DOUBLE) double sample,
+            @SqlType(StandardTypes.DOUBLE) double weight,
+            @SqlType(StandardTypes.VARCHAR) Slice method,
+            @SqlType(StandardTypes.DOUBLE) double min,
+            @SqlType(StandardTypes.DOUBLE) double max)
+    {
+        String requestedMethod = method == null ? null : method.toStringUtf8();
+        if (state.getStrategy() == null) {
+            state.setStrategy(StateSerializer.create(
+                    size, sample, weight, requestedMethod, min, max));
+        }
+        StateStrategy strategy = state.getStrategy();
+        StateSerializer.validate(strategy, requestedMethod);
+        strategy.validateParameters(size, sample, weight, min, max);
+        strategy.add(sample, weight);
+    }
+
+    @InputFunction
+    public static void input(
+            @AggregationState State state,
+            @SqlType(StandardTypes.BIGINT) long size,
+            @SqlType(StandardTypes.DOUBLE) double sample,
+            @SqlType(StandardTypes.DOUBLE) double weight,
+            @SqlType(StandardTypes.VARCHAR) Slice method)
+    {
+        String requestedMethod = method.toStringUtf8();
+        if (state.getStrategy() == null) {
+            state.setStrategy(StateSerializer.create(
```

Comment:
inline create method:

```
            if (requestedMethod.equalsIgnoreCase("fixed_histogram_mle")) {
                state.setStrategy(new FixedHistogramMLEStateStrategy(size, min, max));
            }

            if (requestedMethod.equalsIgnoreCase("fixed_histogram_jacknife")) {
                state.setStrategy(new FixedHistogramJacknifeStateStrategy(size, min, max));
            }
            
            verify(false, "Unsupported method: " + requestedMethod);
```

Reply:
Added to the documentation of `StateSerializer` a string explaining the rationale, since this and a comment from a previous iteration both pertain to the same point. Will be happy to discuss further, of course.